scipy/spatial/transform/_rotation.pyx

```diff
@@ -1273,6 +1345,200 @@ cdef class Rotation:
+        as long as consecutive axes are orthogonal:
+
+        >>> e1 = [2, 0, 0]
+        >>> e2 = [0, 1, 0]
+        >>> e3 = [1, 0, 1]
+        >>> axes = [e1, e2, e3]
+        >>> r = R.from_davenport(axes, 'extrinsic', [90, 45, 30], degrees=True)
+        >>> r.as_quat().shape
+        (4,)
+
+        """
+        if order in ['e', 'extrinsic']:
+            extrinsic = True
+        elif order in ['i', 'intrinsic']:
+            extrinsic = False
+        else:
+            raise ValueError("order should be 'e'/'extrinsic' for extrinsic "
+                             "sequences or 'i'/'intrinsic' for intrinsic "
+                             "sequences, got {}".format(order))
+
+        axes = np.asarray(axes)
+        if axes.ndim == 1:
+            axes = axes.reshape([1, 3])
+
+        num_axes = len(axes)
+
+        for axis in axes:
+            if len(axis) != 3:
+                raise ValueError("Axes must be vectors of length 3.")
+
+        if num_axes < 1 or num_axes > 3:
+            raise ValueError("Expected up to 3 axes, got {}".format(num_axes))
+
+        # normalize axes
+        norm = np.repeat(np.linalg.norm(axes, axis=1), 3)
+        axes = axes / norm.reshape(num_axes, 3)
+
+        if (num_axes > 1 and abs(np.dot(axes[0], axes[1])) >= 1e-7 or 
+            num_axes > 2 and abs(np.dot(axes[1], axes[2])) >= 1e-7):
+            raise ValueError("Consecutive axes must be orthogonal.")
+            
+        angles = np.asarray(angles, dtype=float)
+        if degrees:
+            angles = np.deg2rad(angles)
+
+        is_single = False
+        # Prepare angles to have shape (num_rot, num_axes)
+        if num_axes == 1:
+            if angles.ndim == 0:
+                # (1, 1)
+                angles = angles.reshape((1, 1))
+                is_single = True
+            elif angles.ndim == 1:
+                # (N, 1)
+                angles = angles[:, None]
+            elif angles.ndim == 2 and angles.shape[-1] != 1:
+                raise ValueError("Expected `angles` parameter to have shape "
+                                "(N, 1), got {}.".format(angles.shape))
+            elif angles.ndim > 2:
+                raise ValueError("Expected float, 1D array, or 2D array for "
+                                "parameter `angles` corresponding to `seq`, "
+                                "got shape {}.".format(angles.shape))
+        else:  # 2 or 3 axes
+            if angles.ndim not in [1, 2] or angles.shape[-1] != num_axes:
+                raise ValueError("Expected `angles` to be at most "
+                                "2-dimensional with width equal to number "
+                                "of axes specified, got {} for shape".format(
+                                angles.shape))
+
+            if angles.ndim == 1:
+                # (1, num_axes)
+                angles = angles[None, :]
+                is_single = True
+
+        # By now angles should have shape (num_rot, num_axes)
```

Comment:
Do you mean that the code above guarantees this?

In this case I believe just `assert` will be more idiomatic.

Reply:
What do you mean by that? An assert here would block the code though, wouldn't it?

This is inspired by how the original code for Euler functions was written, in order to accept an angles array of shape `(N, )` as if it where `(1, N)`

Reply:
I simply mean that if the false condition means a bug in the code it's more idiomatic to use assert rather than the exception. "Use exceptions to inform users and other libraries and asserts to check ourselves" (something like that).

Reply:
So is it the case of an additional meaningful check or just extra safety to underline the properties of `angles` at that point?

Reply:
I am actually not sure if this is needed or not, which is why I didn't touch it: just copied from the `from_euler` implementation

Reply:
As I can see the two blocks of code are identical, starting from `angles = np.asarray(angles, dtype=float)`.

Can you refactor it into a separate function like `check_angles` (which returns prepared and checked angles)?

Reply:
Done, created a new `_format_angles` function